substrate/state-machine/src/lib.rs

```diff
@@ -40,70 +41,49 @@ mod testing;
 
 pub use testing::TestExternalities;
 pub use ext::Ext;
-
-/// Updates to be committed to the state.
-pub type Update = (Vec<u8>, Vec<u8>);
-
-// in-memory section of the state.
-#[derive(Debug, PartialEq, Default, Clone)]
-struct MemoryState {
-	storage: HashMap<Vec<u8>, Vec<u8>>,
-}
-
-impl MemoryState {
-	fn storage(&self, key: &[u8]) -> Option<&[u8]> {
-		self.storage.get(key).map(|v| &v[..])
-	}
-
-	fn set_storage(&mut self, key: Vec<u8>, val: Vec<u8>) {
-		self.storage.insert(key, val);
-	}
-
-	fn update<I>(&mut self, changes: I) where I: IntoIterator<Item=Update> {
-		for (key, val) in changes {
-			if val.is_empty() {
-				self.storage.remove(&key);
-			} else {
-				self.storage.insert(key, val);
-			}
-		}
-	}
-}
+pub use backend::Backend;
 
 /// The overlayed changes to state to be queried on top of the backend.
 ///
 /// A transaction shares all prospective changes within an inner overlay
 /// that can be cleared.
 #[derive(Default)]
 pub struct OverlayedChanges {
-	prospective: MemoryState,
-	committed: MemoryState,
+	prospective: HashMap<Vec<u8>, Option<Vec<u8>>>,
+	committed: HashMap<Vec<u8>, Option<Vec<u8>>>,
 }
 
 impl OverlayedChanges {
-	fn storage(&self, key: &[u8]) -> Option<&[u8]> {
-		self.prospective.storage(key)
-			.or_else(|| self.committed.storage(key))
-			.and_then(|v| if v.is_empty() { None } else { Some(v) })
+	/// Returns a double-Option: None if the key is unknown (i.e. and the query should be refered
+	/// to the backend); Some(None) if the key has been deleted. Some(Some(...)) for a key whose
+	/// value has been set.
+	pub fn storage(&self, key: &[u8]) -> Option<Option<&[u8]>> {
```

Comment:
really should be another enum with three variants.

Reply:
considered it but don't agree. it really is a hierarchy.